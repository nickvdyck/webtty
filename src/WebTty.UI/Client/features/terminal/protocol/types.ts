import {
    OpenNewTabRequest,
    OpenNewTabReply,
    ResizeTabRequest,
    SendInputRequest,
    OpenOutputRequest,
    OutputEvent,
} from "@webtty/messages"

class UnknownMessage {
    public type: string
    public payload: unknown

    constructor(type: string, payload: unknown) {
        this.type = type
        this.payload = payload
    }

    public toJSON(): unknown {
        return {}
    }
}

type Messages =
    | OpenNewTabRequest
    | OpenNewTabReply
    | ResizeTabRequest
    | SendInputRequest
    | OpenOutputRequest
    | OutputEvent
    | UnknownMessage

export { UnknownMessage, Messages }
