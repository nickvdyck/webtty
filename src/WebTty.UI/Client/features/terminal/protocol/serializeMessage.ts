import { encode } from "@msgpack/msgpack"
import {
    OpenNewTabRequest,
    OpenNewTabReply,
    ResizeTabRequest,
    SendInputRequest,
    OpenOutputRequest,
    OutputEvent,
} from "@webtty/messages"

import BinaryMessageFormatter from "common/utils/BinaryMessageFormatter"
import { Messages } from "./types"

const getName = (message: unknown): string => {
    if (message instanceof OpenNewTabRequest) return "OpenNewTabRequest"
    if (message instanceof OpenNewTabReply) return "OpenNewTabReply"
    if (message instanceof ResizeTabRequest) return "ResizeTabRequest"
    if (message instanceof SendInputRequest) return "SendInputRequest"
    if (message instanceof OpenOutputRequest) return "OpenOutputRequest"
    if (message instanceof OutputEvent) return "OutputEvent"

    throw new Error("Unknown message!")
}

const serializeMessage = (message: Messages): ArrayBuffer => {
    const payload = encode([getName(message), message.toJSON()])
    return BinaryMessageFormatter.write(payload)
}

export default serializeMessage
