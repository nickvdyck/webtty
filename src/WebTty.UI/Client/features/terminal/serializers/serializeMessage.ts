import { encode } from "@msgpack/msgpack"
import {
    OpenNewTabRequest,
    ResizeTabMessage,
    StdInputRequest,
} from "@webtty/messages"

import BinaryMessageFormatter from "common/BinaryMessageFormatter"
import Messages from "./Messages"

const getName = (message: unknown): string => {
    if (message instanceof OpenNewTabRequest) return "OpenNewTabRequest"
    if (message instanceof ResizeTabMessage) return "ResizeTabMessage"
    if (message instanceof StdInputRequest) return "StdInputRequest"

    return ""
}

const serializeMessage = (message: Messages): ArrayBuffer => {
    const payload = encode([getName(message), message.toJSON()]).slice()
    const data = BinaryMessageFormatter.write(payload)
    return data
}

export default serializeMessage
