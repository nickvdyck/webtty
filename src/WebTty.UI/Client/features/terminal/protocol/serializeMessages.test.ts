import serializeMessage from "./serializeMessage"
import { Messages } from "./types"
import {
    OpenNewTabRequest,
    OpenNewTabReply,
    ResizeTabRequest,
    SendInputRequest,
    OpenOutputRequest,
    OutputEvent,
} from "@webtty/messages"

describe("serializeMessage", () => {
    it.each`
        message
        ${new OpenNewTabRequest()}
        ${new OpenNewTabReply({ id: "123" })}
        ${new ResizeTabRequest({ tabId: "1", cols: 1, rows: 1 })}
        ${new SendInputRequest({ tabId: "1", payload: "h" })}
        ${new OpenOutputRequest({ tabId: "123" })}
        ${new OutputEvent({ tabId: "123", data: [1, 2, 3] })}
    `(
        "serializes a message into an array buffer",
        ({ message }: { message: Messages; expected: ArrayBuffer }): void => {
            // Given, When
            const buffer = serializeMessage(message)

            // Then
            expect(new Uint8Array(buffer)).toMatchSnapshot()
        },
    )

    it("throws an error trying to serialize an unknown message", () => {
        // Given
        const message: any = {}

        // When, Then
        expect(() => serializeMessage(message)).toThrow("Unknown message!")
    })
})
