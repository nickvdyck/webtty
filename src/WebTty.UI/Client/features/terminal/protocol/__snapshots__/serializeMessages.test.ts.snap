// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serializeMessage serializes a message into an array buffer 1`] = `
Uint8Array [
  27,
  146,
  177,
  79,
  112,
  101,
  110,
  78,
  101,
  119,
  84,
  97,
  98,
  82,
  101,
  113,
  117,
  101,
  115,
  116,
  129,
  165,
  84,
  105,
  116,
  108,
  101,
  192,
]
`;

exports[`serializeMessage serializes a message into an array buffer 2`] = `
Uint8Array [
  25,
  146,
  175,
  79,
  112,
  101,
  110,
  78,
  101,
  119,
  84,
  97,
  98,
  82,
  101,
  112,
  108,
  121,
  129,
  162,
  73,
  100,
  163,
  49,
  50,
  51,
]
`;

exports[`serializeMessage serializes a message into an array buffer 3`] = `
Uint8Array [
  39,
  146,
  176,
  82,
  101,
  115,
  105,
  122,
  101,
  84,
  97,
  98,
  82,
  101,
  113,
  117,
  101,
  115,
  116,
  131,
  165,
  84,
  97,
  98,
  73,
  100,
  161,
  49,
  164,
  67,
  111,
  108,
  115,
  1,
  164,
  82,
  111,
  119,
  115,
  1,
]
`;

exports[`serializeMessage serializes a message into an array buffer 4`] = `
Uint8Array [
  37,
  146,
  176,
  83,
  101,
  110,
  100,
  73,
  110,
  112,
  117,
  116,
  82,
  101,
  113,
  117,
  101,
  115,
  116,
  130,
  165,
  84,
  97,
  98,
  73,
  100,
  161,
  49,
  167,
  80,
  97,
  121,
  108,
  111,
  97,
  100,
  161,
  104,
]
`;

exports[`serializeMessage serializes a message into an array buffer 5`] = `
Uint8Array [
  30,
  146,
  177,
  79,
  112,
  101,
  110,
  79,
  117,
  116,
  112,
  117,
  116,
  82,
  101,
  113,
  117,
  101,
  115,
  116,
  129,
  165,
  84,
  97,
  98,
  73,
  100,
  163,
  49,
  50,
  51,
]
`;

exports[`serializeMessage serializes a message into an array buffer 6`] = `
Uint8Array [
  33,
  146,
  171,
  79,
  117,
  116,
  112,
  117,
  116,
  69,
  118,
  101,
  110,
  116,
  130,
  165,
  84,
  97,
  98,
  73,
  100,
  163,
  49,
  50,
  51,
  164,
  68,
  97,
  116,
  97,
  147,
  1,
  2,
  3,
]
`;
